src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1188,30 +1192,88 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
-
-#if HAVE_FCOPYFILE
-    // If fcopyfile is available (OS X), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
-    // Copy data and metadata.
-    return fcopyfile(inFd, outFd, NULL, COPYFILE_ALL) == 0 ? 0 : -1;
-#else
-    // Get the stats on the source file.
+    int outFd;
     int ret;
     struct stat_ sourceStat;
-    bool copied = false;
-#if HAVE_SENDFILE_4
-    // If sendfile is available (Linux), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
+
     while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
     if (ret != 0)
     {
         return -1;
     }
 
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
+            return -1;
+        }
+
+        if (!overwrite)
+        {
+            errno = EEXIST;
+            return -1;
+        }
+
+#if HAVE_CLONEFILE
+        // For clonefile we need to unlink the destination file first but we need to
+        // check permission first to ensure we don't try to unlink read-only file
+        if (access(destPath, W_OK) != 0)
+        {
+            errno = EACCES;
```

Comment:
In this case, should we instead just skip clonefile and try to do the open/read/write loop?  open may fail in a similar way, but it seems less prone to race conditions.

Reply:
The `access` call would still be necessary because the permissions for opening and unlinking files are different. If we don't bail out early then `unlink` will happily remove the read-only file and `open` will succeed.

Reply:
Sorry, what I wrote wasn't clear / what I intended.  What I meant was, if `access` fails, rather than error'ing out, should we fall back to the open/read/write path?

Reply:
I cannot think of any scenario where `access` would fail and `open` would succeed except for a race conditon where someone deletes the file between the `stat` and `access` calls. I suppose we can do the fallback if you feel it's worth the trouble.

Reply:
Ok.